infrastructures/infrastructure-common/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AbstractAddonInfrastructure.java

```diff
@@ -244,29 +233,26 @@ public Void handle() {
      * @param terminateInstanceIfEmpty whether the instance termination will 
      *                                 be requested to the cloud provider
      */
+    @SuppressWarnings("unchecked")
     protected void unregisterNodeAndRemoveInstanceIfNeeded(final String instanceId, final String nodeName,
             final String infrastructureId, final boolean terminateInstanceIfEmpty) {
-        setPersistedInfraVariable(new PersistedInfraVariablesHandler<Void>() {
-            @Override
-            @SuppressWarnings("unchecked")
-            public Void handle() {
-                // first read from the runtime variables map
-                nodesPerInstance = (Map<String, Set<String>>) persistedInfraVariables.get(NODES_PER_INSTANCES_KEY);
-                // make modifications to the nodesPerInstance map
-                nodesPerInstance.get(instanceId).remove(nodeName);
-                logger.info("Removed node : " + nodeName);
-                if (nodesPerInstance.get(instanceId).isEmpty()) {
-                    if (terminateInstanceIfEmpty) {
-                        connectorIaasController.terminateInstance(infrastructureId, instanceId);
-                        logger.info("Instance terminated: " + instanceId);
-                    }
-                    nodesPerInstance.remove(instanceId);
-                    logger.info("Removed instance : " + instanceId);
+        setPersistedInfraVariable(() -> {
+            // first read from the runtime variables map
+            nodesPerInstance = (Map<String, Set<String>>) persistedInfraVariables.get(NODES_PER_INSTANCES_KEY);
+            // make modifications to the nodesPerInstance map
+            nodesPerInstance.get(instanceId).remove(nodeName);
```

Comment:
maybe it is better to check if `nodeName` existed

Reply:
Yes I applied. Thanks.